src/Roslyn.Diagnostics.Analyzers/Core/ImportingConstructorShouldBeObsolete.cs

```diff
@@ -0,0 +1,182 @@
+                return;
+            }
+
+            foreach (var constructor in namedType.Constructors)
+            {
+                if (constructor.IsImplicitlyDeclared)
+                {
+                    continue;
+                }
+
+                var constructorAttributes = constructor.GetAttributes();
+                AttributeData importingConstructorAttributeData = null;
+                foreach (var attributeData in constructorAttributes)
+                {
+                    if (attributeData.AttributeClass.DerivesFrom(importingConstructorAttribute))
+                    {
+                        importingConstructorAttributeData = attributeData;
+                        break;
+                    }
+                }
+
+                if (importingConstructorAttributeData is null)
+                {
+                    // This constructor is not marked [ImportingConstructor]
+                    continue;
+                }
+
+                var foundObsoleteAttribute = false;
+                foreach (var attributeData in constructorAttributes)
+                {
+                    if (!attributeData.AttributeClass.Equals(obsoleteAttribute))
+                    {
+                        continue;
+                    }
+
+                    foundObsoleteAttribute = true;
+                    if (attributeData.ConstructorArguments.Length != 2)
+                    {
+                        if (attributeData.ConstructorArguments.Length == 0)
+                        {
+                            // '{0}' is MEF-exported and should have a single importing constructor of the correct form
+                            context.ReportDiagnostic(Diagnostic.Create(Rule, attributeData.ApplicationSyntaxReference.GetSyntax().GetLocation(), ScenarioProperties.MissingDescription, namedType.Name));
+                            break;
+                        }
+                        else if (attributeData.ConstructorArguments.Length == 1)
+                        {
+                            // '{0}' is MEF-exported and should have a single importing constructor of the correct form
+                            context.ReportDiagnostic(Diagnostic.Create(Rule, attributeData.ApplicationSyntaxReference.GetSyntax().GetLocation(), ScenarioProperties.MissingError, namedType.Name));
+                            break;
+                        }
+                        else
+                        {
+                            // '{0}' is MEF-exported and should have a single importing constructor of the correct form
+                            context.ReportDiagnostic(Diagnostic.Create(Rule, attributeData.ApplicationSyntaxReference.GetSyntax().GetLocation(), namedType.Name));
+                            break;
+                        }
+                    }
+
+                    if (!Equals(attributeData.ConstructorArguments[0].Value, "This exported object must be obtained through the MEF export provider."))
```

Comment:
Why do we care about the message being the same? Can it change in the future to have new alternate forms which breaks this analyzer? I understand the check below for the second argument, but unsure about this check.

Reply:
The code fix for this updates an existing attribute to reference `MefConstruction.ImportingConstructorMessage`.